src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncEHTests.cs

```diff
@@ -1241,6 +1478,718 @@ public static void Main()
+                                int V_2)
+                  IL_0000:  ldc.i4.0
+                  IL_0001:  stloc.0
+                  IL_0002:  ldnull
+                  IL_0003:  stloc.1
+                  .try
+                  {
+                    IL_0004:  call       "System.Threading.Tasks.Task<int> Test.F()"
+                    IL_0009:  call       "int System.Runtime.CompilerServices.AsyncHelpers.Await<int>(System.Threading.Tasks.Task<int>)"
+                    IL_000e:  stloc.0
+                    IL_000f:  ldloca.s   V_0
+                    IL_0011:  constrained. "int"
+                    IL_0017:  callvirt   "string object.ToString()"
+                    IL_001c:  newobj     "System.Exception..ctor(string)"
+                    IL_0021:  throw
+                  }
+                  catch object
+                  {
+                    IL_0022:  stloc.1
+                    IL_0023:  leave.s    IL_0025
+                  }
+                  IL_0025:  ldloc.0
+                  IL_0026:  call       "System.Threading.Tasks.Task<int> Test.F()"
+                  IL_002b:  call       "int System.Runtime.CompilerServices.AsyncHelpers.Await<int>(System.Threading.Tasks.Task<int>)"
+                  IL_0030:  stloc.2
+                  IL_0031:  ldloc.2
+                  IL_0032:  add
+                  IL_0033:  stloc.0
+                  IL_0034:  ldloc.1
+                  IL_0035:  brfalse.s  IL_004c
+                  IL_0037:  ldloc.1
+                  IL_0038:  isinst     "System.Exception"
+                  IL_003d:  dup
+                  IL_003e:  brtrue.s   IL_0042
+                  IL_0040:  ldloc.1
+                  IL_0041:  throw
+                  IL_0042:  call       "System.Runtime.ExceptionServices.ExceptionDispatchInfo System.Runtime.ExceptionServices.ExceptionDispatchInfo.Capture(System.Exception)"
+                  IL_0047:  callvirt   "void System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()"
+                  IL_004c:  ldnull
+                  IL_004d:  throw
+                }
+                """);
+        }
+
+        [Fact]
+        public void AsyncInFinally006_AsyncVoid()
+        {
+            var source = """
+                using System;
+                using System.Threading;
+                using System.Threading.Tasks;
+                class Test
+                {
+                    static async Task<int> F()
+                    {
+                        return 2;
+                    }
+                    static async void G(SemaphoreSlim semaphore)
+                    {
+                        int x = 0;
+                        try
+                        {
+                            x = await F();
+                        }
+                        finally
+                        {
+                            x += await F();
+                            Console.WriteLine(x);
+                            semaphore.Release();
+                        }
```

Comment:
>} [](http://example.com/codeflow?start=24&length=1)

Is there a test with a user code after the finally?

Reply:
Not sure if we have an explicit test, but I can add one here.